### PR TITLE
Implementation docs

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: implementation.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+1. Create a class, _V_, that will hold all the data for a “row” in the incoming
+data. It must be a class and not an interface.
+1. Create a parameterless method in _V_ for every variable to be exposed. The
+method names must be valid Shemsu names (lowercase with underscores) and
+decorated with `@Export` annotations with the correct type signature. All
+methods must return `boolean`, `long`, `String`, `Instant`, `Set`, or `Tuple`
+(and `Set` and `Tuple` may only contain more of the same).
+1. Create a new interface, _R_, extending `InputRepository<`_V_`>`.
+1. Create a new class that extends `BaseInputFormatDefinition<`_V_`,`_R_`> and
```
Comment (morgantaschuk) [inline implementation.md]: Rogue `

File: implementation.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+When binding an instance, a `MethodHandles.Lookup` object must be provided.
+There are two ways to obtain one
+[`publicLookup()`](https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/MethodHandles.html#publicLookup--),
+which can only bind to public methods, and
+[`lookup()`](https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/MethodHandles.html#lookup--),
+which can bind to any method that is visible from the caller.
+
+### Actions
+Actions are the most likely thing that must be added to Shesmu, but they are
```
Comment (morgantaschuk) [inline implementation.md]: This isn't true anymore.

File: implementation.md [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+To create an action:
+
+1. Create a class _A_ that extends `Action`.
+1. It must provide a unique _type_ name to the superconstructor that will be available for searching via the REST API.
+1. Override all the methods for the desired behaviour as described above.
+1. Create a class _AR_ that implements `ActionRepository`.
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `ActionRepository` class.
+1. Create a class _AD_ that extends `ActionDefinition`.
+1. In the constructor of _AD_, call the superconstructor with `("_json type name_`", Type.getType(`_A_`.class), Stream.of(`_parameters_`);`
```
Comment (morgantaschuk) [inline implementation.md]: Another rogue `

File: implementation.md (hunk 2 above)
Comment (apmasell) [reply implementation.md]: No? I think of someone were to pick this up off the shelf, it probably isn't going to work with their workflow engine, even if we collect a bunch of those.

Comment (morgantaschuk) [reply implementation.md]: In the README you say that all kinds of plugins are likely going to need to be implemented for someone to use it locally, so it's not the most likely thing. Maybe the most common thing after deployment?

**Comment by apmasell** [comment]
These docs are unrelated to those docs, so I was working in different branches.

File: implementation.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+plugin interfaces avoid as much of the JVM internals as possible, but some is
+required.
+
+There are plenty of tutorials on JVM internals, but the [DZone bytecode
+tutorial](https://dzone.com/articles/nominalized-adjectives-as-names-for-decorators)
+is one of the better ones.
+
+Bytecode is written using [ASM's
+`GeneratorAdapter`](https://static.javadoc.io/org.ow2.asm/asm/5.2/org/objectweb/asm/commons/GeneratorAdapter.html)
+which provides a methods for writing bytecode that is slightly more high-level
```
Comment (callunity) [inline implementation.md]: `a method` or `methods`?

File: implementation.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+1. The Java type system
+1. The JVM type system
+1. The ASM type system
+1. The Shesmu type system
+1. The JSON type system
+
+Shesmu's type system describe the types of every type that can be used in an
+olive. Because `Type` is used by the ASM library, Shesmu's types are
+represented by `Imyhat` objects (this is an Ancient Egyptian for _mould of
```
Comment (callunity) [inline implementation.md]: `an Ancient Egyptian word for`?

File: implementation.md [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+complicated to most complicated.
+
+### Throttler
+Throttlers temporarily block actions from starting. They can block actions by
+service names provided by plugins and these names are arbitrary strings. To
+create a new throttler:
+
+1. Create a class which implements the `Throttler` interface.
+1. Annotate this class with `@MetaInfServices`.
+1. Make sure the class as a no-argument constructor.
```
Comment (callunity) [inline implementation.md]: `has`?

File: implementation.md [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+1. Create a new interface, _R_, extending `InputRepository<`_V_`>`.
+1. Create a new class that extends `BaseInputFormatDefinition<`_V_`,`_R_`>` and
+provides those types to the constructor as well as a name that will be used in
+the `Input` instruction. This class must be annotated with
+`@MetaInfServices(InputFormatDefinition)`.
+1. Create new classes that provide data which implement _R_ and are annotated
+with `@MetaInfServices(`_R_`)`.
+
+### Constants
+A constant is a value that can be generate with no input. It need not actually
```
Comment (callunity) [inline implementation.md]: `generated`

File: implementation.md [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+There are a number of convience methods to create constants for values. If
+something more complicated is desired, extend the `Constant` class and write
+arbitrary bytecode in the `load` method. The code must not change any value on
+the stack and must leave exactly one value of the correct type on the stack.
+
+### Functions
+A function is a transformation of input data. It matches with a call to a
+method. Since there is no user-defined error-handling, these functions should
+not throw. Also, since Shesmu has no null values, they should not return null
+on an error.
```
Comment (callunity) [inline implementation.md]: `or`?

File: implementation.md [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+To provide functions:
+
+1. Create a class taht implements `FunctionRepository`.
```
Comment (callunity) [inline implementation.md]: `that`

File: implementation.md [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `FunctionRepository` class.
+1. Return a stream of `FunctionDefintion` objects for the functions to be created.
+
+There are a few methods to generate a `FunctionDefinition`:
```
Comment (callunity) [inline implementation.md]: `ways` instead of `methods`? You do use `method` technically elsewhere in the document.

File: implementation.md [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+To create an action:
+
+1. Create a class _A_ that extends `Action`.
+1. It must provide a unique _type_ name to the superconstructor that will be available for searching via the REST API.
+1. Override all the methods for the desired behaviour as described above.
+1. Create a class _AR_ that implements `ActionRepository`.
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `ActionRepository` class.
+1. Create a class _AD_ that extends `ActionDefinition`.
+1. In the constructor of _AD_, call the superconstructor with `("_json type name_`", Type.getType(`_A_`.class)`, Stream.of(`_parameters_`);`
```
Comment (callunity) [inline implementation.md]: The ` after _json type name_ looks unmatched

File: implementation.md (hunk 2 above)
Comment (apmasell) [reply implementation.md]: So I did! :dizzy_face:

**Comment by apmasell** [comment]
Fixed.

**Pull request merged**